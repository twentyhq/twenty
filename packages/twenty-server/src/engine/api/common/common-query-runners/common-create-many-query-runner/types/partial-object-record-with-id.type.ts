import { type ObjectRecord } from 'twenty-shared/types';

export type PartialObjectRecordWithId = Partial<ObjectRecord> & { id: string };
