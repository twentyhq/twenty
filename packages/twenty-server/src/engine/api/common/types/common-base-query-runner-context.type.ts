import { type ObjectLiteral } from 'typeorm';

import { type CommonQueryRunnerOptions } from 'src/engine/api/common/interfaces/common-query-runner-options.interface';
import { type ResolverArgs } from 'src/engine/api/graphql/workspace-resolver-builder/interfaces/workspace-resolvers-builder.interface';

import { type CommonSelectedFieldsResult } from 'src/engine/api/common/types/common-selected-fields-result.type';
import { type WorkspaceDataSource } from 'src/engine/twenty-orm/datasource/workspace.datasource';
import { type WorkspaceRepository } from 'src/engine/twenty-orm/repository/workspace.repository';

export type CommonBaseQueryRunnerContext<Input extends ResolverArgs> = {
  args: Input;
  options: CommonQueryRunnerOptions;
  workspaceDataSource: WorkspaceDataSource;
  repository: WorkspaceRepository<ObjectLiteral>;
  selectedFieldsResult: CommonSelectedFieldsResult;
  isExecutedByApiKey: boolean;
  roleId?: string;
  shouldBypassPermissionChecks: boolean;
};
