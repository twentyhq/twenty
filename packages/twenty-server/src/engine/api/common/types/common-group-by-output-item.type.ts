type AggregateValues = {
  [key: string]: number;
};

type GroupByDimensionValues = {
  groupByDimensionValues: string[];
};

export type CommonGroupByOutputItem = AggregateValues & GroupByDimensionValues;
