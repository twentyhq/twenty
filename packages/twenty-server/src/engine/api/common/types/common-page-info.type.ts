import { type PageInfo } from 'src/engine/api/rest/core/interfaces/rest-api-base.handler';

export type CommonPageInfo = {
  hasNextPage: NonNullable<PageInfo['hasNextPage']>;
  hasPreviousPage: NonNullable<PageInfo['hasPreviousPage']>;
  startCursor: string | null;
  endCursor: string | null;
};
