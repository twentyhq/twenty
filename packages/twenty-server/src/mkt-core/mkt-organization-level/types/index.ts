export * from './organization-level.types';
export * from './permissions.type';
