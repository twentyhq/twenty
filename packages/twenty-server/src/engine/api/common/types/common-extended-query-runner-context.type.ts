import { type ObjectLiteral } from 'typeorm';

import { type WorkspaceAuthContext } from 'src/engine/api/common/interfaces/workspace-auth-context.interface';

import { type CommonBaseQueryRunnerContext } from 'src/engine/api/common/types/common-base-query-runner-context.type';
import { type GraphqlQueryParser } from 'src/engine/api/graphql/graphql-query-runner/graphql-query-parsers/graphql-query.parser';
import { type WorkspaceDataSource } from 'src/engine/twenty-orm/datasource/workspace.datasource';
import { type WorkspaceRepository } from 'src/engine/twenty-orm/repository/workspace.repository';
import { type RolePermissionConfig } from 'src/engine/twenty-orm/types/role-permission-config';

export type CommonExtendedQueryRunnerContext = Omit<
  CommonBaseQueryRunnerContext,
  'authContext'
> & {
  authContext: WorkspaceAuthContext;
  rolePermissionConfig: RolePermissionConfig;
  repository: WorkspaceRepository<ObjectLiteral>;
  commonQueryParser: GraphqlQueryParser;
  workspaceDataSource: WorkspaceDataSource;
};
