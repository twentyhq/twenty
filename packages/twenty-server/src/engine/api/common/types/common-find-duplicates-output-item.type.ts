import { type ObjectRecord } from 'twenty-shared/types';

export type CommonFindDuplicatesOutputItem = {
  records: ObjectRecord[];
  totalCount: number;
  hasNextPage: boolean;
  hasPreviousPage: boolean;
  startCursor: string | null;
  endCursor: string | null;
};
