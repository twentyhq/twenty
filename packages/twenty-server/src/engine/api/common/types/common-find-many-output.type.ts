import { type ObjectRecord } from 'twenty-shared/types';

import { type CommonPageInfo } from 'src/engine/api/common/types/common-page-info.type';
import { type CommonSelectedFieldsResult } from 'src/engine/api/common/types/common-selected-fields-result.type';

export type CommonFindManyOutput = {
  records: ObjectRecord[];
  aggregatedValues: Record<string, number>;
  totalCount: number;
  pageInfo: CommonPageInfo;
  selectedFieldsResult: CommonSelectedFieldsResult;
};
