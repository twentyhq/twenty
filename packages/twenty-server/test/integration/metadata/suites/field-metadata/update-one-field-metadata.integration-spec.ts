import { createOneFieldMetadata } from 'test/integration/metadata/suites/field-metadata/utils/create-one-field-metadata.util';
import { updateOneFieldMetadata } from 'test/integration/metadata/suites/field-metadata/utils/update-one-field-metadata.util';
import {
  LISTING_NAME_PLURAL,
  LISTING_NAME_SINGULAR,
} from 'test/integration/metadata/suites/object-metadata/constants/test-object-names.constant';
import { createOneObjectMetadata } from 'test/integration/metadata/suites/object-metadata/utils/create-one-object-metadata.util';
import { deleteOneObjectMetadata } from 'test/integration/metadata/suites/object-metadata/utils/delete-one-object-metadata.util';
import { FieldMetadataType } from 'twenty-shared/types';

describe('updateOne', () => {
  describe('FieldMetadataService name/label sync', () => {
    let listingObjectId = '';
    let testFieldId = '';

    beforeEach(async () => {
      const { data } = await createOneObjectMetadata({
        input: {
          labelSingular: LISTING_NAME_SINGULAR,
          labelPlural: LISTING_NAME_PLURAL,
          nameSingular: LISTING_NAME_SINGULAR,
          namePlural: LISTING_NAME_PLURAL,
          icon: 'IconBuildingSkyscraper',
          isLabelSyncedWithName: true,
        },
      });

      listingObjectId = data.createOneObject.id;

      const { data: createdFieldMetadata } = await createOneFieldMetadata({
        input: {
          objectMetadataId: listingObjectId,
          type: FieldMetadataType.TEXT,
          name: 'testName',
          label: 'Test name',
          isLabelSyncedWithName: true,
        },
      });

      testFieldId = createdFieldMetadata.createOneField.id;
    });
    afterEach(async () => {
      await deleteOneObjectMetadata({
        input: { idToDelete: listingObjectId },
      });
    });

    it('should update a field name and label when they are synced correctly', async () => {
      // Arrange
      const updateFieldInput = {
        name: 'newName',
        label: 'New name',
        isLabelSyncedWithName: true,
      };

      // Act
      const { data } = await updateOneFieldMetadata({
        input: { idToUpdate: testFieldId, updatePayload: updateFieldInput },
        gqlFields: `
            id
            name
            label
            isLabelSyncedWithName
        `,
      });

      // Assert
      expect(data.updateOneField.name).toBe('newName');
    });

    it('should update a field name and label when they are not synced correctly and labelSync is false', async () => {
      // Arrange
      const updateFieldInput = {
        name: 'differentName',
        label: 'New name',
        isLabelSyncedWithName: false,
      };

      // Act
      const { data } = await updateOneFieldMetadata({
        input: { idToUpdate: testFieldId, updatePayload: updateFieldInput },
        gqlFields: `
              id
              name
              label
              isLabelSyncedWithName
          `,
      });

      // Assert
      expect(data.updateOneField.name).toBe('differentName');
    });

    it('should not update a field name if it is not synced correctly with label and labelSync is true', async () => {
      // Arrange
      const updateFieldInput = {
        name: 'newName',
        isLabelSyncedWithName: true,
      };

      // Act
      const { errors } = await updateOneFieldMetadata({
        input: { idToUpdate: testFieldId, updatePayload: updateFieldInput },
        gqlFields: `
              id
              name
              label
              isLabelSyncedWithName
          `,
        expectToFail: true,
      });

      // Assert
      expect(errors[0].message).toBe(
        'Name is not synced with label. Expected name: "testName", got newName',
      );
    });
  });
});
