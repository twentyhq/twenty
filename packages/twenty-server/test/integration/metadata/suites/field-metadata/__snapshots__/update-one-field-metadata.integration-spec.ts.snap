// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`updateOne failing update should not update a field name if it is not synced correctly with label and labelSync is true 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
      "errors": {
        "index": [],
        "objectMetadata": [
          {
            "errors": [
              {
                "code": "FIELD_MUTATION_NOT_ALLOWED",
                "message": "Name is not synced with label.",
                "userFriendlyMessage": "Updated field name is not synced with label",
              },
            ],
            "flatEntityMinimalInformation": {
              "id": Any<String>,
              "name": "newName",
              "objectMetadataId": Any<String>,
            },
            "type": "update_field",
          },
        ],
        "view": [],
        "viewField": [],
      },
      "message": "Validation failed for 0 object(s) and 0 field(s)",
      "summary": {
        "invalidIndex": 0,
        "invalidObjectMetadata": 0,
        "invalidView": 0,
        "invalidViewField": 0,
        "totalErrors": 0,
      },
      "userFriendlyMessage": "Validation failed for 0 object(s) and 0 field(s)",
    },
    "message": "Multiple validation errors occurred while updating field",
    "name": "GraphQLError",
  },
]
`;

exports[`updateOne failing update should throw if the field name is not available because of other field with the same name 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
      "errors": {
        "index": [],
        "objectMetadata": [
          {
            "errors": [
              {
                "code": "NOT_AVAILABLE",
                "message": "Name "otherTestName" is not available as it is already used by another field",
                "userFriendlyMessage": "Name "otherTestName" is not available as it is already used by another field",
                "value": "otherTestName",
              },
              {
                "code": "FIELD_MUTATION_NOT_ALLOWED",
                "message": "Name is not synced with label.",
                "userFriendlyMessage": "Updated field name is not synced with label",
              },
            ],
            "flatEntityMinimalInformation": {
              "id": Any<String>,
              "name": "otherTestName",
              "objectMetadataId": Any<String>,
            },
            "type": "update_field",
          },
        ],
        "view": [],
        "viewField": [],
      },
      "message": "Validation failed for 0 object(s) and 0 field(s)",
      "summary": {
        "invalidIndex": 0,
        "invalidObjectMetadata": 0,
        "invalidView": 0,
        "invalidViewField": 0,
        "totalErrors": 0,
      },
      "userFriendlyMessage": "Validation failed for 0 object(s) and 0 field(s)",
    },
    "message": "Multiple validation errors occurred while updating field",
    "name": "GraphQLError",
  },
]
`;
